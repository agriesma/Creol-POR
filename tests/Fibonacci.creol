class Fibonacci(n:Integer)
begin
  op init == skip

  op run == var res: Integer; t!fib(n); await t?; t?(res)

  op fib(in n:Integer; out f:Integer) ==
    var a:Integer; var b:Integer;
    if n = 0 then f := 1 else
      if n = 1 then f := 1 else
  	s!fib(n - 1); s?(a); t!fib(n - 2); t?(b); f := a + b
      fi
    fi
with Any
  op fibonacci(in n:Integer; out f:Integer) == t!fib(n); t?(f)
end

class FibTester(f: Fibonacci, n: Integer)
begin
	op init == skip
	op run ==
	 	var res:Integer;
 		t!f.fibonacci(n);
 		t?(res);
end
