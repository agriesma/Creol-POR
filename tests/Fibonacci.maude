load interpreter
mod PROGRAM is
pr INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Fibonacci" : Cl | Inh: noInh, Par: "n", Att: "n" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: "res" |-> null, Code: "t" ! "this" . "fib" ( "n" ) ; ( await ( "t" ?? ) ) ; ( "t" ? ( "res" ) )  ; return ( emp ) > *
  < "fib" : Mtdname | Param: "n", Latt: "n" |-> null , "f" |-> null , "a" |-> null , "b" |-> null, Code: if "=" ( "n" # int(0) ) th "f" ::= int(1) el if "=" ( "n" # int(1) ) th "f" ::= int(1) el "s" ! "this" . "fib" ( "-" ( "n" # int(1) ) ) ; ( "s" ? ( "a" ) )  ; "t" ! "this" . "fib" ( "-" ( "n" # int(2) ) ) ; ( "t" ? ( "b" ) )  ; "f" ::= "+" ( "a" # "b" ) fi fi ; return ( "f" ) > *
  < "fibonacci" : Mtdname | Param: "n", Latt: "n" |-> null , "f" |-> null, Code: "t" ! "this" . "fib" ( "n" ) ; ( "t" ? ( "f" ) )  ; return ( "f" ) >, Ocnt: 0 >

< "FibTester" : Cl | Inh: noInh, Par: "f" , "n", Att: "f" |-> null , "n" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: "res" |-> null, Code: "f" . "fibonacci" ( "n" ; "res" ) ; return ( emp ) >, Ocnt: 0 >

.
endm
