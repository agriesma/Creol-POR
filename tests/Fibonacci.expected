==========================================
reduce in PROGRAM : init .
result Configuration: 
< 'FibTester : Cl | Inh: noInh,Par: 'f,,'n,Att: 'f |-> null, 'n |-> null,Mtds:
< 'init : Mtdname | Param: noAid,Latt: noSubst,Code: skip ; return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: 'res |-> null,Code: 't ! 'f . 'fibonacci(
    'n) ; 't ?('res) ; return(emp) >,Ocnt: 0 >
 
< 'Fibonacci : Cl | Inh: noInh,Par: 'n,Att: 'n |-> null,Mtds:
< 'fib : Mtdname | Param: 'n,Latt: 'a |-> null, 'b |-> null, 'f |-> null, 'n
    |-> null,Code: if 'equal[['n # int(0)]] th 'f ::= int(1) el if 'equal[['n #
    int(1)]] th 'f ::= int(1) el 's ! 'this . 'fib('minus[['n # int(1)]]) ; 's
    ?('a) ; 't ! 'this . 'fib('minus[['n # int(2)]]) ; 't ?('b) ; 'f ::= (
    'plus[['a # 'b]]) fi fi ; return('f) > *
< 'fibonacci : Mtdname | Param: 'n,Latt: 'f |-> null, 'n |-> null,Code: 't !
    'this . 'fib('n) ; 't ?('f) ; return('f) > *
< 'init : Mtdname | Param: noAid,Latt: noSubst,Code: skip ; return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: 'res |-> null,Code: 't ! 'this . 'fib('n)
    ; await ('t ??) ; 't ?('res) ; return(emp) >,Ocnt: 0 >

Bye.
